packages/server/modules/core/services/ratelimiter.ts

```diff
@@ -0,0 +1,317 @@
+      limitCount: getIntFromEnv('RATELIMIT_BURST_GET_OBJECTS_STREAMID_OBJECTID', '200'),
+      duration: 1 * TIME.minute
+    }
+  },
+  'GET /objects/:streamId/:objectId/single': {
+    regularOptions: {
+      limitCount: getIntFromEnv('RATELIMIT_GET_OBJECTS_STREAMID_OBJECTID_SINGLE', '3'),
+      duration: 1 * TIME.second
+    },
+    burstOptions: {
+      limitCount: getIntFromEnv(
+        'RATELIMIT_BURST_GET_OBJECTS_STREAMID_OBJECTID_SINGLE',
+        '200'
+      ),
+      duration: 1 * TIME.minute
+    }
+  },
+  'POST /graphql': {
+    regularOptions: {
+      limitCount: getIntFromEnv('RATELIMIT_POST_GRAPHQL', '50'),
+      duration: 1 * TIME.second
+    },
+    burstOptions: {
+      limitCount: getIntFromEnv('RATELIMIT_BURST_POST_GRAPHQL', '200'),
+      duration: 1 * TIME.minute
+    }
+  }
+}
+
+export const sendRateLimitResponse = (
+  res: express.Response,
+  rateLimitBreached: RateLimitBreached
+): express.Response => {
+  res.setHeader('Retry-After', rateLimitBreached.msBeforeNext / 1000)
+  res.removeHeader('X-RateLimit-Remaining')
+  res.setHeader(
+    'X-RateLimit-Reset',
+    new Date(Date.now() + rateLimitBreached.msBeforeNext).toISOString()
+  )
+  res.setHeader('X-Speckle-Meditation', 'https://http.cat/429')
+  return res.status(429).send({
+    err: 'You are sending too many requests. You have been rate limited. Please try again later.'
+  })
+}
+
+export const getActionForPath = (path: string, verb: string): RateLimitAction => {
+  try {
+    const maybeAction = `${verb} ${path}` as keyof typeof RateLimitAction
+    const action = RateLimitAction[maybeAction]
+    if (!action) throw new Error()
+    return action
+  } catch {
+    return RateLimitAction.ALL_REQUESTS
+  }
+}
+
+export const getSourceFromRequest = (req: express.Request): string => {
+  let source: string | null =
+    ((req as RequestWithAuthContext)?.context?.userId as string) ||
+    getIpFromRequest(req)
+
+  if (!source) source = 'unknown'
+  return source
+}
+
+export const createRateLimiterMiddleware = (
+  rateLimiterMapping: RateLimiterMapping = RATE_LIMITERS
+) => {
+  return async (
+    req: express.Request,
+    res: express.Response,
+    next: express.NextFunction
+  ) => {
+    if (isTestEnv()) return next()
+    const path = req.originalUrl ? req.originalUrl : req.path
+    const action = getActionForPath(path, req.method)
+    const source = getSourceFromRequest(req)
+
+    const rateLimitResult = await getRateLimitResult(action, source, rateLimiterMapping)
+    if (isRateLimitBreached(rateLimitResult)) {
+      return sendRateLimitResponse(res, rateLimitResult)
+    } else {
+      try {
+        res.setHeader('X-RateLimit-Remaining', rateLimitResult.remainingPoints)
+        return next()
+      } catch (err) {
+        if (!(err instanceof RateLimitError)) throw err
+        return sendRateLimitResponse(res, err.rateLimitBreached)
+      }
+    }
+  }
+}
+
+// we need to take the Bursty specific type because its not an Abstract.
```

Comment:
are these comments still applicable? if not, please delete them, if yes convert to a `todo` with some kind of less opaque message that details what needs to be done here